### DISCUSSION TITLE
[Paper Suggestion]: DeepSeek-R1: Incentivizing Reasoning Capability in LLMs via Reinforcement Learning

### DISCUSSION THREAD

**Discussion by m3hrdadfi** [dsai-nlp/core#6] [Paper Suggestions]
### Paper Title

DeepSeek-R1: Incentivizing Reasoning Capability in LLMs via Reinforcement Learning

### Paper Link

https://arxiv.org/abs/2501.12948

### Brief Summary

We introduce our first-generation reasoning models, DeepSeek-R1-Zero and DeepSeek-R1. DeepSeek-R1-Zero, a model trained via large-scale reinforcement learning (RL) without supervised fine-tuning (SFT) as a preliminary step, demonstrates remarkable reasoning capabilities. Through RL, DeepSeek-R1-Zero naturally emerges with numerous powerful and intriguing reasoning behaviors. However, it encounters challenges such as poor readability, and language mixing. To address these issues and further enhance reasoning performance, we introduce DeepSeek-R1, which incorporates multi-stage training and cold-start data before RL. DeepSeek-R1 achieves performance comparable to OpenAI-o1-1217 on reasoning tasks. To support the research community, we open-source DeepSeek-R1-Zero, DeepSeek-R1, and six dense models (1.5B, 7B, 8B, 14B, 32B, 70B) distilled from DeepSeek-R1 based on Qwen and Llama.

### Voting Instructions (do not edit)

```markdown
---
## 🗳️ Voting Instructions
To vote for this paper, **click the ↑ upvote button in the top-left corner of this discussion post**
We'll prioritize papers with the most votes!  

↓ Right here
```

**Comment by lovhag** [comment]
My take on the paper ⬇️

## Questions – What confused you? What would you ask the author?
- How was the evaluation in section 2.2.4 performed? For example, the authors write "One of the most remarkable aspects of this self-evolution is the emergence of sophisticated behaviors as the test-time computation increases. Behaviors such as reflection—where the model revisits and reevaluates its previous steps—and the exploration of alternative approaches to problem-solving arise spontaneously" - how was this measured/quantified?
- Few-shot prompting was found to degrade model performance and the authors encourage using the models zero-shot, describing the problem as opposed to giving examples. Is this a typical finding for RL trained models? Why does it occur? 

## Challenges/Concerns – Anything that seemed weak or unclear?
- More of a debate point that can be skipped: The authors write that "Large Language Models (LLMs) have been undergoing rapid iteration and evolution (Anthropic, 2024; Google, 2024; OpenAI, 2024a), progressively diminishing the gap towards Artificial General Intelligence (AGI)." while some people would argue that LLMs are good at generating strings, while this does not necessarily imply AGI.
- It would have been nice to learn more about the computational cost of the methods/models developed as opposed to relevant baselines (e.g. DeepSeek-V3 trained without RL).

## What the paper did well – Cool ideas, elegant methods, etc.
- Interesting questions asked. E.g. to what extent we can leverage RL only, avoiding SFT.

## Future Research Opportunities – What could be done next?
- The authors struggled with making the model stick to the target language (language mixing), and found that performance degraded when forcing the model to stick to one language. It could be interesting to investigate more "philosophical" research questions related to whether it is easier to solve some problems in a particular language, and if some languages are more beneficial for reasoning (e.g. would a model struggle more in English compared to Chinese?).

**Reply by ricj** [reply]
>  found that performance degraded when forcing the model to stick to one language

There is some previous research showing that CoT can improve end results *even when the reasoning makes no sense*. One idea is that CoT-based models simply have more computation at their disposal, and that one shouldn't take their reasoning literally. https://arxiv.org/pdf/2404.15758

**Reply by ricj** [reply]
Re AGI: I guess from a purely practical point of view, recent systems have more of the "G" since they are quite flexible at performing new tasks on the fly

**Reply by m3hrdadfi** [reply]
> Re AGI: I guess from a purely practical point of view, recent systems have more of the "G" since they are quite flexible at performing new tasks on the fly

I think it's back to the question of whether we have a universal definition for that. I would like to quote Ben Goertzel, who states, 
> AGI = AI that can generalize imaginatively and fluently beyond its training data and its programming.
https://x.com/bengoertzel/status/1853873397132103865

**Comment by nicaudinet** [comment]
I shared a lot of the same questions and sentiments as Lovisa above! Especially around their confident claims about moving towards AGI at the start ... how can you claim you're moving closer to a destination when you don't know what it is?

Anyways, here's a couple more!

## Questions – What confused you? What would you ask the author?
- Why did they use that particular RL method? I recall George saying at the DSAI seminar a couple weeks ago that there was another paper that said that the particular method didn't matter much (didn't find it with a quick google)
- Why did they choose that particular reward model?

## Challenges/Concerns – Anything that seemed weak or unclear?
- I wonder: how does forcing a model to "think" with an emphasis on maths / logical problems affect the model's biases in other domains (for instance, does the model become more utilitarian when asked moral questions?). Do these models become more engineer-like?
- I don't agree / like their split between "reasoning data" and "non-reasoning data" in section 2.3.3 ... seems quite reductive?

## What the paper did well – Cool ideas, elegant methods, etc.
- I liked that they added an "Unsuccessful Attemps" section with an summary of why the methods didn't work
- I thought the paper was generally quite well-organised and readable, with the right amount of detail

## Future Research Opportunities – What could be done next?
- It feels like a waste to just discard the RL trained model to generate the fine-tuning data for the second one ... are there other good uses for it?

**Comment by NicklasOsterbacka** [comment]
### Questions
- R1-Zero "naturally acquires the ability to solve increasingly complex reasoning tasks by leveraging extended test-time computation." How does the content of the CoT evolve as training progresses beyond length? Is CoT length the main contributing factor to the model's reasoning capabilities?
- R1 falls short of V3 in several tasks. Does teaching a model to reason always diminish its capability to do "simple tasks," or is this a consequence of how R1 was trained?

### Challenges/Concerns
- The authors are deliberately coy about the RL reward function, writing that it "mainly consists of two types of rewards" and do not elaborate further. (Maybe it's published in a repo I am unaware of!)

### What the paper did well
- Showing that RL is sufficient to induce reasoning patterns is an extremely interesting result!

### Future Research Opportunities
- Can we use RL to teach a model to make better use of CoT rather than just increasing CoT length as Fig. 3 seems to imply?

**Reply by ricj** [reply]
>  writing that it "mainly consists of two types of rewards" and do not elaborate further.

Much of the technical machinery in the current paper seems to be taken from an earlier paper (https://arxiv.org/pdf/2402.03300). Could the "two types of rewards" be the "Outcome Supervision" and "Process Supervision" described in sec 4.1 in the 2024 paper? Honestly I have just had the time to skim, but can it be that the "Outcome supervision" is a trained reward model and "Process supervision" some way to distribute that reward over tokens?

**Comment by m3hrdadfi** [comment]
**Questions**
- They show that the model can reflect (i.e., "aha moment"), but how often does it do that (all the time)? Can we say that it sometimes reflects unnecessarily, like a hallucinated doubt state?
- Small models outperform even larger ones after distillation but what doesn't survive during distillation? What gets lost during distillation?
- _"We observe that CoT often exhibits language mixing... we introduce a language consistency reward... Although ablation experiments show that such alignment results in a slight degradation in the model’s performance, this reward aligns with human preferences..."_ why language mixing might have been helping?

**Challenges/Concerns**
- CoT and RL help on complex tasks like math and code, where deep thinking is needed. But for tasks that require simple factual recall (e.g., trivia, dates, or definitions), does RL actually help?
- It seems that __more thinking = better performance__. But how much is enough?

**What the paper did well**
- They showed that large-scale reasoning can emerge through pure RL without SFT, reducing dependency on human-labeled data. Although the idea for DeepSeek-R1 was somewhat different, they ran the RL on the SFT version of DeepSeek-v3.

**Future Research Opportunities**
- _"One of the most remarkable aspects of this self-evolution is the emergence of sophisticated behaviors as test-time computation increases. Behaviors such as reflection ..."_ They noted that reflection emerges implicitly as token patterns. Can we detect and quantify reflection in the model's internal activations?

**Comment by dsaynova** [comment]
### Questions – What confused you? What would you ask the author?

- What was the distillation method in section 2.4? Is it simply fine-tuning smaller models with data produced by DeepSeek-R1?
- What would be the effect of the second term in the GRPO loss - the KL divergence with the old (reference) network? Could this be why the language capabilities degrade when RL is applied directly to non-SFT model? What would be the effect of tuning the beta parameter?

### Challenges/Concerns – Anything that seemed weak or unclear?

- I share Nicklas' concern that the reward functions and models are very vague throughout the paper.
- The paper lacks motivation for the proposed methodology for DeepSeek-R1 (beyond the first STF and RL stages). This makes sense for an engineering paper, but makes it hard to evaluate the contributions of each stage/method to model behaviour. 
- The data used is described quite vaguely, and it seems to depend on other models quite a bit. 

### What the paper did well – Cool ideas, elegant methods, etc.

- Interesting methodology for human-less training of reasoning capabilities. 
- Extensive benchmark results (even if I am not familiar with all of them).

### Future Research Opportunities – What could be done next?

- I think it will be interesting to dig into the "emergent" behaviours. 
- If the SFT is only needed to make the output more human-friendly, could we find new reasoning patterns if we study the behaviour of the non-SFT model (discover more efficient ways of solving problems)?

**Reply by m3hrdadfi** [reply]
> * What was the distillation method in section 2.4? Is it simply fine-tuning smaller models with data produced by DeepSeek-R1?

The authors state that: 

"To equip more efficient smaller models with reasoning capabilities like DeepSeek-R1, we directly
fine-tuned open-source models like Qwen ... For distilled models, we apply only SFT and do not include an RL stage, even though
incorporating RL could substantially boost model performance. Our primary goal here ... ", 

Yes, they first ran SFT on 800k samples generated by DeepSeek-R1 and the teacher model used for this distillation process was DeepSeek-R1 itself.

**Reply by m3hrdadfi** [reply]
> * What would be the effect of the second term in the GRPO loss - the KL divergence with the old (reference) network? Could this be why the language capabilities degrade when RL is applied directly to non-SFT model? What would be the effect of tuning the beta parameter?

They didn't mention in their paper that, even they didn't show the effect of the hyperparameters, such as ε and β, either. However, I think it's defined as a penalty term to ensure the new policy does not drift from the reference policy. I remember from Andrew Karpathy's video that he said RL discovers ways to game the model; there are ways to achieve a high reward score even when the inputs were not part of their training set, yet still obtain high scores.

**Reply by dsaynova** [reply]
My second question was not what the second term does in general, but what would be the effect here. That is - is the non-SFT model a good reference point to tie your model to (this is specifically for the DeepSeek-R1-Zero model, which shows language issues related to readability). Since a lot of improvements on the RL methods seem to be adding different ways of constricting the updates, I'm wondering if they can simply be applied like this - maybe in this setting, there are better ways to restrict them.

**Reply by ricj** [reply]
I have interpreted the second term of PPO (and GRPO) to be intended as a regularizer that penalizes the RL-trained model for walking away too much from the basic SFT-trained model
